Durability.md

```diff
@@ -0,0 +1,338 @@
+- Communication between the main thread and logging thread occurs via a bounded, ordered queue.
+- Avoids stalling issues with long running command execution/LUA script execution 
+- Requires coordination between main thread and logging thread about offset progress.
+- Could extend the io threads framework used primarily for networking.
+
+>  Note:
+>
+> * We should start with same thread for end-to-end operation for P0 and then move to further improvements.
+> * Should we consider an approach for initial release to allow the logging component to be available as in-memory instead of fsync it on disk? This mechanism would allow higher throughput and lower latency. Need to assess the risks/recovery time on node failure to perform a full synchronization from the primary and it's affect on failure to reach quorum.
+
+## Topology
+
+Topology management within the durability specification is governed by the RAFT consensus model. Each shard operates as an independent RAFT group, and all membership change such as adding or removing replicas are handled as configuration change entries in the RAFT log. This ensures that updates to the shard topology are serialized, replicated, and committed like any other operation, maintaining a consistent and deterministic view of membership across all nodes.
+
+When a new node joins, the leader initiates a configuration change proposal, replicates it to the quorum, and applies it once committed. Removal follows the same mechanism, guaranteeing that no two conflicting configurations exist concurrently. RAFT’s consensus approach is used during reconfiguration to safely transition between old and new membership sets, avoiding split-brain conditions.These configuration entries are replicated through the same log pipeline as normal commands, ensuring that both data and topology are part of the same durable mechanism.
+
+Here we cover various operation via sequence diagram within a shard.
+
+1. Bootstrap
+2. Node addition
+3. Primary removal
+4. Replica removal
+
+### Bootstrap ([UML code](https://github.com/hpatro/valkey-rfc/blob/durability_hld/assets/UML.md#uml-code-for-bootstrap))
+
+![Bootstrap](https://github.com/hpatro/valkey-rfc/blob/durability_hld/assets/ClusterBootstrap.png)
+
+> Note:
+>
+> * How to discover peer nodes? 
+>   * We can't reuse the `REPLICAOF` mechanism as we intend to piggyback on RAFT's consensus model for leader election.
+>   * Hence, new nodes can discover peers via the `shard-nodes` config which is a list of ip:port address.
+> * When to start the initialization phase? - To be added
+
+### Node addition([UML Code](https://github.com/hpatro/valkey-rfc/blob/durability_hld/assets/UML.md#uml-code-for-node-addition))
+
+![Node Addition](https://github.com/hpatro/valkey-rfc/blob/durability_hld/assets/NodeAddition.png)
+
+### Primary removal ([UML Code](https://github.com/hpatro/valkey-rfc/blob/durability_hld/assets/UML.md#uml-code-for-primary-removal))
```

Comment:
Regarding primary removal, please review section **3.10 Leadership transfer extension** in Ongaro's PhD: https://web.stanford.edu/~ouster/cgi-bin/papers/OngaroPhD.pdf

It is essentially the same coordinated failover algorithm that we already have in Valkey. It would be a pity to ignore it, only to switch to something inferior.

1. The leader pauses writes.
2. The leader waits for the target server to catch up the log index (or replication offset).
3. The leader sends a TimeoutNow request to the target server. This request has the same
effect as the target server’s election timer firing: the target server starts a new election (incrementing its term and becoming a candidate).